File: multitype/src/main/java/com/github/lany192/multitype/adapter/BindingAdapter.java
package com.github.lany192.multitype.adapter;

import android.view.LayoutInflater;
import android.view.ViewGroup;

import androidx.annotation.LayoutRes;
import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.viewbinding.ViewBinding;

import com.chad.library.adapter.base.BaseQuickAdapter;
import com.chad.library.adapter.base.viewholder.BaseViewHolder;

import org.jetbrains.annotations.NotNull;

import java.util.ArrayList;
import java.util.List;

public abstract class BindingAdapter<T extends Object, VB extends ViewBinding> extends BaseQuickAdapter<T, BaseViewHolder> {
    private VB binding;

    public BindingAdapter() {
        super(0, new ArrayList<T>());
    }

    public BindingAdapter(List<T> data) {
        super(0, data);
    }

    private BindingAdapter(@LayoutRes int layoutResId) {
        super(0);
    }

    private BindingAdapter(int layoutResId, @Nullable List<T> data) {
        super(0, data);
    }

    protected abstract VB getBinding(@NonNull LayoutInflater inflater, @NonNull ViewGroup parent);

    @NotNull
    @Override
    protected BaseViewHolder onCreateDefViewHolder(@NotNull ViewGroup parent, int viewType) {
        binding = getBinding(LayoutInflater.from(parent.getContext()), parent);
        return new BaseViewHolder(binding.getRoot());
    }

    @Override
    protected void convert(@NonNull BaseViewHolder viewHolder, T t) {
        bind(binding, t, viewHolder.getBindingAdapterPosition());
    }

    protected abstract void bind(VB binding, T t, int position);
}

File: arch/src/main/java/com/github/lany192/arch/adapter/BindingAdapter.kt
package com.github.lany192.arch.adapter

import android.view.LayoutInflater
import android.view.ViewGroup
import androidx.annotation.LayoutRes
import androidx.viewbinding.ViewBinding
import com.chad.library.adapter.base.BaseQuickAdapter
import com.github.lany192.binding.getBinding
import java.lang.reflect.ParameterizedType

abstract class BindingAdapter<T, VB : ViewBinding> : BaseQuickAdapter<T, BindingHolder<VB>> {

    constructor(data: MutableList<T>) : super(0, data)

    private constructor(@LayoutRes layoutResId: Int) : super(layoutResId)

    private constructor(layoutResId: Int, data: MutableList<T>) : super(layoutResId, data)

    override fun onCreateDefViewHolder(parent: ViewGroup, viewType: Int): BindingHolder<VB> {
        val binding: VB = getClass<VB>(1).getBinding(LayoutInflater.from(parent.context), parent)
        return BindingHolder(binding)
    }

    override fun convert(holder: BindingHolder<VB>, item: T) {
        convert(holder.binding, item, holder.bindingAdapterPosition)
    }

    protected abstract fun convert(binding: VB, item: T, position: Int)

    /**
     * 获取第几个泛型的class
     */
    open fun <T> getClass(index: Int): Class<T> {
        return (javaClass.genericSuperclass as ParameterizedType).actualTypeArguments[index] as Class<T>
    }

}